### ISSUE TITLE
Emulate LIMIT .. WITH TIES using RANK()

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#6197]
jOOQ 3.10 introduces `LIMIT .. WITH TIES` for native support in:

- SQL Server: `TOP n WITH TIES`
- Oracle 12c: `FETCH FIRST n ROWS WITH TIES`

See https://github.com/jOOQ/jOOQ/issues/4900.

We should also be able to emulate the `WITH TIES` clause for all other databases. These are the same:

```sql
-- Native / SQL Standard
SELECT *
FROM t_book
ORDER BY author_id
FETCH FIRST 1 ROW WITH TIES

-- Emulation
SELECT *
FROM (
  SELECT t_book.*, RANK() OVER (ORDER BY author_id) r
  FROM t_book
)
WHERE r = 1
```

We can reuse the logic for SQL Server 2008 `OFFSET` emulation, or Oracle 11g / DB2 `OFFSET .. FETCH` emulation on any database that supports window functions, including PostgreSQL, MySQL 8.0, Firebird, etc.

Note, the real emulation logic is more complicated:

- We mustn't project the `r` column in the outer query
- This means we must rename all the `t_book` columns in the inner query and rename them back to the original name in the outer query
- What happens when we use `DISTINCT` or `UNION`?

**Comment by lukaseder** [comment]
Reopening until we're sure this works with `DISTINCT` and/or `UNION` as well.

**Comment by lukaseder** [comment]
Gaah, @michael-simons, what have you done? :) `WITH TIES` applies to `FETCH` only, not to `OFFSET`. The `OFFSET` part needs to be emulated with `ROW_NUMBER()`, not with `RANK()`

**Comment by michael-simons** [comment]
Sorry :) (Sort of, I would have reacted the same).

**Comment by lukaseder** [comment]
> I would have reacted the same

What? The *aaaargh* part or the "I have to get this right" part? :)

**Comment by michael-simons** [comment]
Like "Sorry, partner, I'll be off for a while, please don't talk to me, keep the kids away, I have to get this right."

**Comment by lukaseder** [comment]
Excellent. Here's a fitting meme, I think:

![image](https://cloud.githubusercontent.com/assets/734593/25898603/17884380-358d-11e7-87f2-8952bb15a0d7.png)

**Comment by lukaseder** [comment]
The `DISTINCT` part will be fixed in another issue: #7201
